privacyidea/api/periodictask.py

```diff
@@ -0,0 +1,182 @@
+# -*- coding: utf-8 -*-
+#  2018-07-09 Friedrich Weber <friedrich.weber@netknights.it>
+#             Initial implementation of periodic tasks
+#
+# This code is free software; you can redistribute it and/or
+# modify it under the terms of the GNU AFFERO GENERAL PUBLIC LICENSE
+# License as published by the Free Software Foundation; either
+# version 3 of the License, or any later version.
+#
+# This code is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU AFFERO GENERAL PUBLIC LICENSE for more details.
+#
+# You should have received a copy of the GNU Affero General Public
+# License along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+
+__doc__ = """These endpoints are used to create, modify and delete
+periodic tasks.
+
+This module is tested in tests/test_api_periodictask.py"""
+
+import json
+import logging
+
+from flask import Blueprint, g, request
+
+from privacyidea.api.lib.prepolicy import prepolicy, check_base_action
+from privacyidea.api.lib.utils import send_result, getParam
+from privacyidea.lib.error import ParameterError
+from privacyidea.lib.policy import ACTION
+from privacyidea.lib.log import log_with
+from privacyidea.lib.periodictask import get_periodic_tasks, set_periodic_task, delete_periodic_task, \
+    enable_periodic_task, get_periodic_task_by_id, get_taskmodule, get_available_taskmodules
+from privacyidea.lib.utils import is_true
+
+log = logging.getLogger(__name__)
+
+periodictask_blueprint = Blueprint('periodictask_blueprint', __name__)
+
+
+def convert_datetimes_to_string(ptask):
+    """
+    Convert the ``last_runs`` timestamps to ISO 8601 strings and return a copy of ``ptask``.
+    :param ptask: periodic task dictionary
+    :return: a new periodic task dictionary
+    """
+    ptask = ptask.copy()
+    ptask['last_runs'] = dict((node, timestamp.isoformat()) for node, timestamp in ptask['last_runs'].items())
+    return ptask
+
+
+@periodictask_blueprint.route('/taskmodules', methods=['GET'])
```

Comment:
Maybe I only realize this now and we have done it wrong in some places. But at least we have done it right e.g. with the endpoint ``/token/``. To my understanding a REST convention would be (and we also did it this way with /user/ and /token/ ) that the endpoint ends with a slash, if the endpoint returns a list.

In this case, the route should be 

   route('/taskmodules/', methods=["GET"])

I know that this is not consistent with e.g. /event/actions/<handlermodule>/, which should end with a slash by this defintion, since it would return actions of a handler module.

Maybe I just want to kick off a discussion. ;-)

Reply:
You're right, ``/taskmodules`` is there because I imitated the events API :-) Honestly I have no idea what the best practice would be here -- I'll read a bit about REST and get back to this.

Reply:
Hm, it seems like people have very different opinions about trailing slashes :-)

[This page](https://restfulapi.net/resource-naming/) (point 2 of "Consistency is the key") says we should not use trailing slashes, and the first comment already says we should always use trailing slashes. [Here](https://apievangelist.com/2015/08/04/how-do-we-handle-that-trailing-slash-in-our-api-endpoints/) are some more opinions.

Personally I think adding a slash makes sense, because we return a collection of task modules.

But thinking about the RESTfulness of the API:
* Shouldn't ``/options/<taskmodule>`` rather be ``/taskmodules/<taskmodule>/options`` then? Because ``/options/`` doesn't identify a resource.
* Aren't ``/enable/<ptaskid>`` and ``/disable/<ptaskid>`` not really RESTful either? ``/enable/`` and ``/disable/`` do not identify resources either :-/

... complex topic! :-)